core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java

```diff
@@ -161,7 +164,16 @@ public boolean incrementNumUnfinishedResponses() {
     @Override
     public boolean canSendRequest() {
         assert responseDecoder != null;
-        return active && !responseDecoder.needsToDisconnectWhenFinished();
+        if (!channel.isActive()) {
+            return false;
+        }
+
+        if (responseDecoder instanceof Http2ResponseDecoder) {
+            final Http2GoAwayHandler goAwayHandler = ((Http2ResponseDecoder) responseDecoder).goAwayHandler();
```

Comment:
For HTTP/2, multiple requests try to use a channel. If one of the requests contains `Connection: close`, we need to deactivate the channel so that subsequent requests don't acquire the channel anymore. `active` is set to `false`.

Meanwhile, in-flight requests that have already acquired the channel try to send messages. If we use `active` to determine if the requests are able to be sent, all in-flight requests are failed with `UnprocessedException`.

As a result, I removed `active` flag from the condition to allow sending inflight requests before a GOAWAY is received or sent for HTTP/2.

Reply:
I see, thanks for the explanation.
I'm still a little unsure whether we don't need to check `!responseDecoder.needsToDisconnectWhenFinished()`.

For instance if `KeepAliveHandler#needsDisconnection() = true` due to `maxConnectionAgeExceeded`, then am I understanding correctly that `HttpSessionHandler#canAcquire` is `true` and the client will keep sending requests?

Reply:
Good point. It does not seem properly handled for HTTP/2. If `maxConnectionAgeExceeded` becomes true, a session is not released to the pool for HTTP/1. But there is no such logic for HTTP/2. I made `canAcquire()` return false when `maxConnectionAgeExceeded` is triggered.

Reply:
>  We shouldn't also send a request when currentNumRequests >= maxNumRequestsPerConnection is true

I believe this change [`a524587` (#4531)](https://github.com/line/armeria/pull/4531/commits/a5245874c0f9c10b5b993ddf2ee2d0250461be98) may cover that case. I didn't want to immediately make `canSendRequest()` false when `currentNumRequests` exceeds `maxNumRequestsPerConnection` that causes a lot of `ClosedSessionException` for HTTP/2.